### PR TITLE
Add the.while

### PR DESCRIPTION
**Pull request by ryanspoone**
## Description

This PR aims to add functionality to loop a function until a desired condition is met. This is helpful for instances where you do not know the end of a loop because it is dependent on the results of a function. 

```js
const the = require('await-the');
let sum = 0;
const condition = previousResult => sum < 10;
const asyncFn = x => {
    sum += x;
    return sum * 10;
}
const result = await the.while(condition, asyncFn, 2);
// will loop while sum < 10, then return the final function result
// loop1: sum is 2
// loop2: sum is 4
// loop3: sum is 6
// loop4: sum is 8
// loop5: sum is 10, returns 100
```

## How Has This Been Tested

Added unit tests for `the.while`.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "await-the",
-    "version": "3.1.4",
+    "version": "3.2.0",
```
Comment (marcusmotill) [inline package.json]: 3.1.5 so you don't need to bump a lot unless you want to

File: test/while.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            await the.while(
+                () => true,
+                () => {
+                    throw new Error(errorMessage);
+                }
+            );
+            assert(false, 'should never get here');
+        } catch (error) {
+            assert(error);
+            assert.strictEqual(error.message, errorMessage);
```
Comment (marcusmotill) [inline test/while.js]: test for the `...args` logic missing? whats the code cov report locally when running npm test

Comment (ryanspoone) [reply test/while.js]: Yeah, that's a good one.

```
---------------|----------|----------|----------|----------|-------------------|
File           |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
---------------|----------|----------|----------|----------|-------------------|
All files      |    98.48 |    81.61 |    95.74 |    98.43 |                   |
 await-the     |      100 |      100 |      100 |      100 |                   |
  index.js     |      100 |      100 |      100 |      100 |                   |
 await-the/lib |    98.34 |    81.61 |    95.74 |    98.28 |                   |
  Limiter.js   |      100 |    92.86 |      100 |      100 |           152,154 |
  all.js       |      100 |      100 |      100 |      100 |                   |
  any.js       |      100 |      100 |      100 |      100 |                   |
  callback.js  |      100 |      100 |      100 |      100 |                   |
  each.js      |      100 |      100 |      100 |      100 |                   |
  map.js       |      100 |       50 |    83.33 |      100 |          27,44,47 |
  mapValues.js |      100 |       50 |    83.33 |      100 |          28,43,46 |
  result.js    |    76.92 |     62.5 |      100 |    76.92 |          38,43,51 |
  retry.js     |      100 |    89.47 |      100 |      100 |             31,32 |
  wait.js      |      100 |       50 |      100 |      100 |                20 |
  while.js     |      100 |      100 |      100 |      100 |                   |
---------------|----------|----------|----------|----------|-------------------|
```

File: lib/while.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                .then(res => {
+                    loop(res);
+                })
+                .catch(reject);
+        };
+        Promise.resolve()
+            .then(loop)
+            .catch(reject);
+    });
+};
```
Comment (marcusmotill) [inline lib/while.js]: how about these bois
```js
module.exports = async (condition, fn, ...args) => {
    if (!condition()) {
        return;
    }

    const loop = async () => {
        const result = await fn(...args);
        if (condition(result)) {
            await loop();
        }
    };

    return await loop();
};
```

Comment (marcusmotill) [reply lib/while.js]: also probably want to support `condition === true` for infinite loops

Comment (marcusmotill) [reply lib/while.js]: wonder if we can get `break;` to work?

Comment (marcusmotill) [reply lib/while.js]: oh and lets make `condition` support promises

Comment (ryanspoone) [reply lib/while.js]: Added all except for `break`.

File: lib/while.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ * @returns {*} The thrown error or the result.
+ */
+module.exports = async (condition, fn, ...args) => {
+    if ((await condition()) !== true) {
+        return;
+    }
+
+    const loop = async () => {
+        const result = await fn(...args);
+        if ((await condition(result)) === true) {
```
Comment (marcusmotill) [inline lib/while.js]: does this need to `=== true` like regaular whiles aren't strict to true just truthy so 
```suggestion
        if ((await condition(result))) {
```

Comment (ryanspoone) [reply lib/while.js]: Is that not what you meant in this comment?

> also probably want to support `condition === true` for infinite loops

https://github.com/olono/await-the/pull/31#discussion_r291341561

**Pull request merged**